#!/bin/bash
#SBATCH -c32
#SBATCH --mem=110g
#SBATCH --gres=lscratch:200
#SBATCH --time=12:0:0

#D15 sample took 8.5 h, 32 cpu and 91 GB.
set -e
module load parallel minimap2/2.24 sambamba/0.8.2
sample=$1
mkdir -p temp_bam bam
find PAG10600 -type f -name "*fastq.gz" | parallel -j 10 -I% --max-args 1 --tag "minimap2 -a -x map-ont -Y -t 3 -R "@RG\\\\tLB:nisc\\\\tID:PAG10600\\\\tSM:$sample" /data/OGL/resources/genomes/NCBI/GRCh38Decoy/genome.mmi % | sambamba sort -u --compression-level 6 --tmpdir=/lscratch/$SLURM_JOB_ID -t 3 -o %.bam <(sambamba view -S -f bam --compression-level 0 -t 3 /dev/stdin)"
find PAG57572 -type f -name "*fastq.gz" | parallel -j 10 -I% --max-args 1 --tag "minimap2 -a -x map-ont -Y -t 3 -R "@RG\\\\tLB:nisc\\\\tID:PAG57572\\\\tSM:$sample" /data/OGL/resources/genomes/NCBI/GRCh38Decoy/genome.mmi % | sambamba sort -u --compression-level 6 --tmpdir=/lscratch/$SLURM_JOB_ID -t 3 -o %.bam <(sambamba view -S -f bam --compression-level 0 -t 3 /dev/stdin)"
find PAG57959 -type f -name "*fastq.gz" | parallel -j 10 -I% --max-args 1 --tag "minimap2 -a -x map-ont -Y -t 3 -R "@RG\\\\tLB:nisc\\\\tID:PAG57959\\\\tSM:$sample" /data/OGL/resources/genomes/NCBI/GRCh38Decoy/genome.mmi % | sambamba sort -u --compression-level 6 --tmpdir=/lscratch/$SLURM_JOB_ID -t 3 -o %.bam <(sambamba view -S -f bam --compression-level 0 -t 3 /dev/stdin)"
find PAG* -type f -name "*.bam" -exec mv {} temp_bam \;
find PAG* -type f -name "*.bam.bai" -exec mv {} temp_bam \;
#The minimap2 index were generated by: minimap2 -d genome.mmi /fdb/igenomes/Homo_sapiens/NCBI/GRCh38Decoy/Sequence/WholeGenomeFasta/genome.fa 

#-x map-ont: saw secondary alignment. Did not see secondary alignment when -x map-ont not set previously. 